src/ui/public/field_editor/lib/copy_field.js

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { has } from 'lodash';
+
+export const copyField = (field, indexPattern, Field) => {
+  const changes = {};
+  const shadowProps = {
+    toActualField: {
+      // bring the shadow copy out of the shadows
```

Comment:
I'm not actually sure what this means. 😄 Could we make this clearer?

Reply:
@cj @bmcconaghy 

Your confusion is warranted! I felt the same way looking at the original code (where this is copied from) 😆 

I added some comments to the top of the function, but will try to explain a little further here. This all has to do with the nature of index pattern `Field` type objects:
https://github.com/elastic/kibana/blob/f0cc432fafd6f7053f32d493aba8b576dcb20539/src/ui/public/index_patterns/_field.js

They are created with custom [property descriptors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#Description) using `ObjDefine`:
https://github.com/elastic/kibana/blob/f0cc432fafd6f7053f32d493aba8b576dcb20539/src/ui/public/utils/obj_define.js

Each property of a `Field` type object could be enumerable/non-enumerable, writable/not writable, configurable/not configurable, and have custom getter and setter. We can't use the original `field` object directly for creating a new field or editing a new field, since we need all the properties to be editable.

A normal copy of `field` (i.e. `const newField = { ...field }`) will only copy enumerable properties and copy each property's descriptors (not writable, etc).

So what this function does is copy `field`'s **property descriptors** into `newFieldProps` (I just renamed this from `shadowProps` for clarity) and modify them so that they are "writable" with a getter/setter that stores and retrieves changes into/from another object (`changes`).

Then it finally returns a new object created from the new property descriptors (`Object.create(null, newFieldProps)`).

When we are ready to save the copied field back into the index pattern, we use `toActualField()` to retrieve an actual `Field` type object, using its original properties and our `changes`.

I wish this wasn't so complex, but simplifying would require changes to `Field` which I don't want to do in the scope of this PR. Hopefully this clears some things up (or maybe not 🙂).